File: data/readable_format/cpp/file485.cpp
#include <iostream>
#include <cassert>

#define SINGLE_PRECISION 1
#define DOUBLE_PRECISION 2

namespace ppm_module_util_gmres {

constexpr int ppm_gmres_param_success = 0;
constexpr int ppm_gmres_param_failure = 1;
constexpr int ppm_gmres_param_maxiter = 2;

// Simulated implementation for single precision
template<int Precision>
int ppm_util_gmres() {
    std::cout << "Called GMRES with precision: " << Precision << std::endl;
    return ppm_gmres_param_success; // Dummy success
}

// Simulated implementation for double precision
template<int Precision>
int ppm_util_gmres_solveupper() {
    std::cout << "Called solve upper with precision: " << Precision << std::endl;
    return ppm_gmres_param_success; // Dummy success
}

} // namespace ppm_module_util_gmres

int main() {
    using namespace ppm_module_util_gmres;

    // Test single precision GMRES
    assert(ppm_util_gmres<SINGLE_PRECISION>() == ppm_gmres_param_success);
    std::cout << "Single precision GMRES test passed." << std::endl;

    // Test double precision GMRES
    assert(ppm_util_gmres<DOUBLE_PRECISION>() == ppm_gmres_param_success);
    std::cout << "Double precision GMRES test passed." << std::endl;

    // Test single precision solve upper
    assert(ppm_util_gmres_solveupper<SINGLE_PRECISION>() == ppm_gmres_param_success);
    std::cout << "Single precision solve upper test passed." << std::endl;

    // Test double precision solve upper
    assert(ppm_util_gmres_solveupper<DOUBLE_PRECISION>() == ppm_gmres_param_success);
    std::cout << "Double precision solve upper test passed." << std::endl;

    return 0;
}

File: data/readable_format/fortran/file485.f90
module ppm_module_util_gmres
   implicit none
   integer, parameter :: ppm_gmres_param_success = 0
   integer, parameter :: ppm_gmres_param_failure = 1
   integer, parameter :: ppm_gmres_param_maxiter = 2

contains

   ! Dummy implementation of single precision GMRES utility function
   function ppm_util_gmres_s() result(res)
       integer :: res
       print *, "Called single precision GMRES utility function."
       res = ppm_gmres_param_success
   end function ppm_util_gmres_s

   ! Dummy implementation of double precision GMRES utility function
   function ppm_util_gmres_d() result(res)
       integer :: res
       print *, "Called double precision GMRES utility function."
       res = ppm_gmres_param_success
   end function ppm_util_gmres_d

   ! Dummy implementation of single precision solve upper function
   function ppm_util_gmres_solveupper_s() result(res)
       integer :: res
       print *, "Called single precision solve upper function."
       res = ppm_gmres_param_success
   end function ppm_util_gmres_solveupper_s

   ! Dummy implementation of double precision solve upper function
   function ppm_util_gmres_solveupper_d() result(res)
       integer :: res
       print *, "Called double precision solve upper function."
       res = ppm_gmres_param_success
   end function ppm_util_gmres_solveupper_d

end module ppm_module_util_gmres

program test_gmres
   use ppm_module_util_gmres
   implicit none

   integer :: result

   ! Test single precision GMRES
   result = ppm_util_gmres_s()
   if (result == ppm_gmres_param_success) then
       print *, "Single precision GMRES test passed."
   else
       print *, "Single precision GMRES test failed."
   endif

   ! Test double precision GMRES
   result = ppm_util_gmres_d()
   if (result == ppm_gmres_param_success) then
       print *, "Double precision GMRES test passed."
   else
       print *, "Double precision GMRES test failed."
   endif

   ! Test single precision solve upper
   result = ppm_util_gmres_solveupper_s()
   if (result == ppm_gmres_param_success) then
       print *, "Single precision solve upper test passed."
   else
       print *, "Single precision solve upper test failed."
   endif

   ! Test double precision solve upper
   result = ppm_util_gmres_solveupper_d()
   if (result == ppm_gmres_param_success) then
       print *, "Double precision solve upper test passed."
   else
       print *, "Double precision solve upper test failed."
   endif

end program test_gmres